### PR TITLE
Provide a way to control dynamic version caching for imported version catalog

### LINKED ISSUE #26009: Provide a way to control dynamic version caching for imported version catalog
### Expected Behavior

I would like to disable caching for version catalog which is defined with dynamic version. With the existing API available for `build.gradle.kts`, it could look like this:

```kotlin
// settings.gradle.kts

dependencyResolutionManagement {
    versionCatalogs {
        create("libs") {
            from("com.example:my-version-catalog:[2.0.0,2.1.0)")
        }
    }
}

configurations {
    named("incomingCatalogForLibs0") {
        resolutionStrategy.cacheDynamicVersionsFor(0, "seconds")
    }
}
```

### Current Behavior (optional)

When a new version of the version catalog is published, it is not automatically resolved in the projects (unless explicitly specifying the `--refresh-dependencies` parameter).

### Context

We have a shared version catalog that is used in all our services. When doing minor version bumps, we release the next patch version of the catalog (like from `2.0.18` to `2.0.19`) and we would like to make our services automatically use it when configured with a dynamic version like `[2.0.0,2.1.0)`. Now we have to pass the `--refresh-dependencies` parameter which has global scope and makes our build slower.

### PR DESCRIPTION
**Pull request by radoslaw-panuszewski**
Fixes #26009

### Context
Users who use version catalog with dynamic version can disable/configure the cache.

In this PR I propose to reuse the existing API for configuring dynamic version caching:
```kotlin
dependencyResolutionManagement {
    configurations.all {
        cacheDynamiVersionsFor(0, "seconds")
    }
}
```

#### Advantages:
* ✅  we reuse existing API and follow the same convention as for regular configurations
* ✅  it provides more flexibility for the user

#### Disadvantages:
* ⛔️ we can't explicitly use it like that:
```kotlin
dependencyResolutionManagement {
    configurations {
        named("incomingCatalogForLibs0") {
            cacheDynamiVersionsFor(0, "seconds")
        }
    }
}
```
because the `incomingCatalogForLibs0` configuration is created ad hoc later and can't be found during the settings file evaluation.
* ⛔️ due to the above limitation, it will always affect all configurations used internally by the dependency resolution management

### Alternative API

Maybe we could think of some less general API, tailored to this specific use case. For example:
```kotlin
dependencyResolutionManagement {
    cacheVersionCatalogDynamicVersionsFor(0, "seconds")
}
```
but I'm not sure about it.

Please let me know what you think 😉 

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by radoslaw-panuszewski** [comment]
@jvandort could you review this PR? 🙂

File: subprojects/docs/src/docs/userguide/dep-man/03-controlling-transitive-dependencies/platforms.adoc [hunk 1]
```diff
@@ -430,6 +430,26 @@ In the example above, any dependency which was using the `groovy` version as ref
...
+
+You can specify a dynamic version for your version catalog import:
+
+.Importing version catalog with the dynamic version
+====
+include::sample[dir="snippets/dependencyManagement/catalogs-javaPlatformCatalog/kotlin/consumer",files="settings.gradle.kts[tags=dynamic_version]"]
+include::sample[dir="snippets/dependencyManagement/catalogs-javaPlatformCatalog/groovy/consumer",files="settings.gradle[tags=dynamic_version]"]
+====
+
+By default, Gradle caches dynamic versions for 24 hours. To change how long Gradle will cache the resolved version of the version catalog, use:
```
Comment (lkasso) [inline subprojects/docs/src/docs/userguide/dep-man/03-controlling-transitive-dependencies/platforms.adoc:444]:
```suggestion
By default, Gradle caches dynamic versions for 24 hours. 
To change how long Gradle will cache the resolved version of the version catalog, use:
```

**Review by lkasso — APPROVED** [review]

File: subprojects/docs/src/docs/userguide/dep-man/03-controlling-transitive-dependencies/platforms.adoc (hunk 1 above)
Comment (lkasso) [reply subprojects/docs/src/docs/userguide/dep-man/03-controlling-transitive-dependencies/platforms.adoc:444]: Docs are approved with a very minor change requested.

**Review by jvandort — COMMENTED** [review]
I've discussed with the team and I do not think we want to follow-through this this approach. We are working on moving away from the `Condfiguration` API entirely so we do not want to add new APIs that expose it. 

In general, I am a bit skeptical about needing to control the caching for dynamic versions in the version catalog. It seems counterintuitive to use dynamic versions in the version catalog, as the catalog is really meant to control and document a static set of versions used by a build

**Comment by radoslaw-panuszewski** [comment]
> In general, I am a bit skeptical about needing to control the caching for dynamic versions in the version catalog. It seems counterintuitive to use dynamic versions in the version catalog, as the catalog is really meant to control and document a static set of versions used by a build

In our use case, we have a shared version catalog (for our team only) and we use it in a number of services. Our goal is to keep them up-to-date without much manual work. We tried dependabot and it didn't work for us, because of the overwhelming number of PRs it opened to each of our services (and we have >10 of them).

This shared version catalog approach works really well, but we don't want to remember about bumping it every time we change something in the service. For minor version updates in the catalog, we release a patch version of it, that's why we have our services configured like `version-catalog:[2.0.0,2.1.0)`. The default caching mechanism is a problem for us because we have to add the `--refresh-dependencies` flag for the new version catalog to be picked up (if we don't want to wait 24 hours for it). If the update in version catalog can cause any breaking changes in our services, we release minor or major version of it. This won't be picked up automatically until we explicitly change the version range, for example to `[2.1.0,2.2.0)`.

I find it counter-intuitive that Gradle _allows_ to use dynamic versions for the version catalog, but it _disallows_ to configure it. For me, it looks like a design flaw.

**Comment by jvandort** [comment]
There is indeed a gap in the API here, however we definitely do not want to expose the `Configuration` API here. 

Dynamic versions in general are tricky since they essentially make build reproducibility impossible without dependency locking. Whenever using dynamic versions, [dependency locking](https://docs.gradle.org/current/userguide/dependency_locking.html) should also be used to ensure reproducibility. Otherwise builds can unexpectedly change behavior as new versions are published and caches expire. 

In this particular case, disabling the dynamic version cache negatively affects performance by requiring Gradle to check dynamic versions every build, while `--refresh-dependencies` completely obliterates performance altogether. Dependency locking solves this by allowing you to choose when to update dynamic versions with `--write-locks`, while also making builds reproducible. 

It is clear that there is a hole in the API here, but to me it seems we should plug that by either
1) Disabling dynamic versions at all for version catalogs
2) Allowing users to configure dependency locking for version catalogs

It might even make sense to add warnings/deprecate using dynamic versions without dependency locking altogether. 

I do not see disabling caching as the right solution here. 

In any case, we definitely do not want to expose the `Configuration` API here, as we are working towards moving away from it. As such, I'm going to close this PR and we should move this conversation over to the linked issue.

**Comment by radoslaw-panuszewski** [comment]
Hey @jvandort, thanks for the explanation 🙂 [Let's continue the discussion](https://github.com/gradle/gradle/issues/26009#issuecomment-1789527716) in the linked issue then.

**Pull request closed without merging**